### DISCUSSION TITLE
Range based for loops and integer overflow

### DISCUSSION THREAD

**Discussion by zagortenay333** [odin-lang/Odin#986] [Q&A]
While making range based loops for my own compiler I noticed an awkward case. The following loop will run forever in odin:

```odin
package main

main :: proc() {
    x : u8 = 0;
    y : u8 = 255;
    for i in x..y { }
}
```
It's because the loop terminates when x is strictly larger than y, but that's not possible here.

Is this intended behavior? I'll probably leave it like this in my own toy compiler since it's easier to handle.

EDIT
------

I opened an issue for this: https://github.com/odin-lang/Odin/issues/990

**Comment by ThisDevDane** [comment]
@gingerBill Huh, that doesn't seem intended at all? It indeed just spins around and around.

**Reply by gingerBill** [reply]
This was intended behaviour but I have now made Odin follow this more "intuitive" behaviour by adding an explicit check to these kinds of for loops, meaning an extra branch may be required (assuming it does not get optimized out).

**Comment by Kelimion** [comment]
Seems to me that it's treated as <256, so it'll overflow to zero. 0..0, infinite loop.

**Reply by zagortenay333** [reply]
It's not 0..0. What happens is: 0..255   ->   254..255   ->   255..255   ->   0..255   ->   ....

**Reply by ghost** [reply]
Ooooh, that's a good point! I think Kelimion you're saying that the compiler is using <256 instead of <= 255?
This bug is pretty weird, lol.

**Reply by Kelimion** [reply]
> Ooooh, that's a good point! I think Kelimion you're saying that the compiler is using <256 instead of <= 255?
> This bug is pretty weird, lol.

That was my guess at the time based on the wrapping behavior. It's since been investigated more deeply.

**Comment by gingerBill** [comment]
I have implemented the more "intuitive" behaviour now into the Odin compiler: https://github.com/odin-lang/Odin/issues/990

Previously, `for i in x .. y {}` was implemented like this:
```odin
for i := x; i <= y; i += 1 {}
```
Which will cause an infinite loop if `y` is equal to the maximum value of that integer type.

The new behaviour acts like this:
```odin
for i := x; i <= y; /**/ {
    // body
   if i == y {
       break;
   }
   i += 1;
}
```

If `y` is known at compile time by the backend, then that branch can be removed. However if `y` is not know at compile time (especially if y is a procedure call or something), then that branch must exist.

**Reply by gingerBill** [reply]
It should also be noted that internal integer promotion may not be relied upon, especially if the integer is the maximum integer type already.

**Reply by gingerBill** [reply]
`for i in x ..< y {}` does not need any special treatment

**Reply by zagortenay333** [reply]
@gingerBill The loop you show above does not match what I suggested. My implementation only performs one check per iteration, yours does two: `i<=y` and `i==y`.
In my implementation you jump unconditionally to the start of the loop body after `i==y` fails.

**Reply by zagortenay333** [reply]
To be precise, it performs two checks for the first iteration, and only 1 check for the subsequent iterations.
